[UI automation] support app update scenarios

somalaya opened the pull request:
**What** : Currently we install the app and perform UI automation testcase run on one particular version of the app. But we do not have testcases that test update scenarios. 
For ex : AcquireToken interactive is called on an app and the next update of the broker app is available. The user updates the app and acquireToken silent is called. To test whether this scenario is successful, I added update support.

**How** : Added update() method to IApp to update with the given installer. By default the app will be updated using the  playstore version if the installer is not explicitly provided.  In each of the broker apps like BrokerMicrosoftAuthentictor, BrokerCompanyPortal, BrokerHost and AzureSampleApp - Added the name of the localApk which will be used for updating the app.

Note : Internally update method added in App.java calls adb install only. But this will still be an app update as adb install -r is the closest you get, that is actually an update as it keeps the database and stored preferences. If you uninstall/re-install both the app database and preferences is deleted. 

**Testing** : Added testcases to test the acquireTokenSilent call after updating the app in https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1640

somalaya committed:
update scenarios support

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -108,6 +127,21 @@ public void launch() {
+
+        // the app is just installed, first run should be handled
+        // this value can (should) be changed to false by child classes as appropriate
+        shouldHandleFirstRun = true;
```
Is this actually required for update?

somalaya replied:
I tested the update scenario and this is not required as first run flag is stored in storage in authenticator and this is not reset on update. Hence removing this piece of code

shahzaibj commented:
You mentioned that `-r` flag is needed to re-install. I don't see that in the PR though. Is our code somewhere already passing the `-r` flag to ADB? Can you point to that code?

somalaya committed:
update method added

somalaya committed:
added local path for adb update

iamgusain commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/broker/AbstractTestBroker.java:
```diff
@@ -63,6 +63,9 @@ public abstract class AbstractTestBroker extends App implements ITestBroker {
     public final static IAppInstaller DEFAULT_BROKER_APP_INSTALLER = BuildConfig.INSTALL_SOURCE_LOCAL_APK
             .equalsIgnoreCase(BuildConfig.BROKER_INSTALL_SOURCE)
             ? new LocalApkInstaller() : new PlayStore();
+    public final static IAppInstaller DEFAULT_BROKER_APP_UPDATE_INSTALLER = BuildConfig.INSTALL_SOURCE_LOCAL_APK
```
Should this be managed via a different BuildConfig?
In case we want to first install playstore version and then update using local Apk

shahzaibj replied:
+1

somalaya replied:
Added a buildConfig for update source as well

iamgusain reviewed (approved):
Left minor comment, can be addressed separately

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -76,6 +81,8 @@ public App(@NonNull final String packageName, @NonNull final String appName) {
         this.appInstaller = appInstaller;
         this.packageName = packageName;
+        // update installer is PlayStore by default
+        this.updateAppInstaller = new PlayStore();
```
I think tis is fine for now, but it would actually make more sense for LocalApk installer to be the default here as usually we would be testing update scenarios with a side loaded APK

somalaya replied:
Changed it to use LocalApkInstaller instead

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -108,6 +127,17 @@ public void launch() {
+    public void update() {
+        if (updateAppInstaller instanceof LocalApkInstaller && !TextUtils.isEmpty(localUpdateApkFileName)) {
+            Logger.i(TAG, "Updating the " + this.appName + " from local apk..");
+            AdbShellUtils.updatePackage(LocalApkInstaller.LOCAL_APK_PATH_PREFIX + localApkFileName, "-t", "-r", "-d");
```
Why not put this as a method on the `LocalApkInstaller`? I mean create a new method in that file called as `updateApp` - similar to what you've done for PlayStore, and then over here you can probably just invoke `updateAppInstaller.updateApp(identifier)`

somalaya replied:
addressed this in f9ecbb25a67397feee33a63807075d46e1b84038

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/installer/PlayStore.java:
```diff
@@ -187,4 +187,49 @@ public void installApp(@NonNull final String searchHint) {
         installAppFromMarketPage();
     }
 
+    public void updateApp(@NonNull final String searchHint) {
```
Consider my other suggestion and put this as interface method also on the `IAppInstaller`

somalaya replied:
addressed in f9ecbb25a67397feee33a63807075d46e1b84038

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/utils/AdbShellUtils.java:
```diff
@@ -99,6 +99,28 @@ public static void installPackage(@NonNull final String packageName, @NonNull St
+     * @param packageName the name of the package to update
+     * @param flags       the flags to use during for update of the app
+     */
+    public static void updatePackage(@NonNull final String packageName, @NonNull String... flags) {
```
Looks like it is doing the same thing as the `installPackage` method. Why do we need this method?

somalaya replied:
removed this, I added this method to add few differences but in the end removed them.

somalaya committed:
installing old apk first and then using latest apk for update

somalaya committed:
update broker source using build.config property

shahzaibj commented on uiautomationutilities/build.gradle:
```diff
@@ -42,10 +42,14 @@ android {
 
     def final installSourcePlayStore = "PlayStore"
     def final installSourceLocalApk = "LocalApk"
+    def final updateSourcePlayStore = "PlayStore"
+    def final updateSourceLocalApk = "LocalApk"
```
There is duplication of constants now. I think you can reduce these to just 2 as opposed to 4:

- appSourcePlayStore = "PlayStore"
- appSourceLocalApk = "LocalApk"

shahzaibj commented on uiautomationutilities/build.gradle:
```diff
@@ -81,11 +97,15 @@ android {
         buildConfigField("String", "WORD_APP_INSTALL_SOURCE", "\"$wordAppInstallationSource\"")
         buildConfigField("String", "INSTALL_SOURCE_PLAY_STORE", "\"$installSourcePlayStore\"")
         buildConfigField("String", "INSTALL_SOURCE_LOCAL_APK", "\"$installSourceLocalApk\"")
+        buildConfigField("String", "UPDATE_SOURCE_PLAY_STORE", "\"$updateSourcePlayStore\"")
+        buildConfigField("String", "UPDATE_SOURCE_LOCAL_APK", "\"$updateSourceLocalApk\"")
```
See my recommendation above about reducing duplicated constants

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -84,20 +91,43 @@ public App(@NonNull final String packageName,
-        } else {
-            Logger.i(TAG, "Installing the " + this.appName + " from Play Store..");
-            appInstaller.installApp(packageName);
+        if (localOldApkFileName != null) {
+            // To ensure to install the old apk first in update scenarios
+            if (appInstaller instanceof LocalApkInstaller && !TextUtils.isEmpty(localOldApkFileName)) {
+                Logger.i(TAG, "Installing the " + this.appName + " from local apk..");
+                appInstaller.installApp(localOldApkFileName);
+            } else {
+                Logger.i(TAG, "Installing the " + this.appName + " from Play Store..");
+                appInstaller.installApp(packageName);
+            }
+        }
+        else {
+            //TODO: make it build time configurable to specify the installer that should be used.
+            // Ideally we can specify different installers on app basis
+            if (appInstaller instanceof LocalApkInstaller && !TextUtils.isEmpty(localApkFileName)) {
+                Logger.i(TAG, "Installing the " + this.appName + " from local apk..");
+                appInstaller.installApp(localApkFileName);
+            } else {
+                Logger.i(TAG, "Installing the " + this.appName + " from Play Store..");
+                appInstaller.installApp(packageName);
+            }
```
I think you can consider reducing the number of `if/else` blocks here. As a hint, null check and `TextUtils.isEmpty` check are similar with the latter being a more comprehensive check

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -116,6 +146,19 @@ public void launch() {
+    @Override
+    public void update() {
+        String appHint;
+        if (updateAppInstaller instanceof LocalApkInstaller && !TextUtils.isEmpty(localApkFileName)) {
```
What if the initial install is also done using the same `localApkFileName`?

somalaya replied:
This is not possible in update scenarios. I have pushed changes in this (https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1640/files) PR to make sure that in case of update scenarios - We extend AbstractMsalBrokerUpdateTest in all update scenarios to first install old apk and then install the latest one when update function is called

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/AzureSampleApp.java:
```diff
@@ -47,12 +47,21 @@ public class AzureSampleApp extends App {
         localApkFileName = AZURE_SAMPLE_APK;
     }
 
+    public AzureSampleApp(final boolean isOldApk) {
+        super(AZURE_SAMPLE_PACKAGE_NAME, AZURE_SAMPLE_APP_NAME, new LocalApkInstaller());
+        localApkFileName = AZURE_SAMPLE_APK;
+        if (isOldApk) {
+            localOldApkFileName = OLD_AZURE_SAMPLE_APK;
+        }
+    }
```
There is no javadoc and/or comments here and it is hard to tell what this method is achieving

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/broker/BrokerCompanyPortal.java:
```diff
@@ -77,6 +78,14 @@ public BrokerCompanyPortal(@NonNull final IAppInstaller appInstaller) {
         localApkFileName = COMPANY_PORTAL_APK;
     }
 
+    public BrokerCompanyPortal(final boolean isOldApk) {
```
I think it is difficult for the readers of this code to figure out what would happen if we pass a true vs false to this constructor. It seems that we've got some boolean here that is governing how to pick the APK file name when doing install.

I think rather than having `oldApkFileName` and dynamically setting it based on a boolean, I think instead we should have clearly defined `localApkFileNameFreshInstall` and `localApkFileNameUpdate` - something like that and then you don't need a booleans in the constructor to figure these out

shahzaibj replied:
The above feedback to all the places these patterns are used in this PR

somalaya replied:
The reason I did not add localApkFileNameFreshInstall and localApkFileNameUpdate is because when we say update - we will actually be installing the latest apk (one that is picked from the latest pipeline) and so, when we run update scenarios, we first install an oldApk based on the testcase that is running. 

I could have kept static files like you mentioned. But in that case, I would end up installing localApkFileNameFreshInstall apks everytime for the first install - here it would become tricky what a fresh install means as we cannot differentiate between freshInstall in normal testcase and freshInstall in update testcase (in this case freshInstall means oldApk install)

To fix the ambiguity, I can add some comments to make it clear

shahzaibj replied:
I think I kind of get what you're saying, but still not fully clear. Comments in the code would definitely help. Let's also sync on this offline next week so I can get a better understanding of the flow here.

somalaya committed:
removed duplicate vars

somalaya committed:
added constructor to accept old and update apks

somalaya committed:
added constructor for update and addressed comments

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/installer/PlayStore.java:
```diff
@@ -187,4 +187,49 @@ public void installApp(@NonNull final String searchHint) {
         installAppFromMarketPage();
     }
 
+    public void updateApp(@NonNull final String searchHint) {
```
Is this an overriden method? If yes, then let's add override annotation

somalaya replied:
Done

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/installer/PlayStore.java:
```diff
@@ -187,4 +187,49 @@ public void installApp(@NonNull final String searchHint) {
         installAppFromMarketPage();
     }
 
+    public void updateApp(@NonNull final String searchHint) {
+        if (isStringPackageName(searchHint)) {
+            launchMarketPageForPackage(searchHint);
+        } else {
+            searchAppOnGooglePlay(searchHint);
+            selectGooglePlayAppFromAppName();
+        }
+
+        updateAppFromMarketPage();
+    }
+
+    private void updateAppFromMarketPage() {
+        Logger.i(TAG, "Update App From Market Page..");
+        try {
+            updateAppFromMarketPageInternal();
+        } catch (final UiObjectNotFoundException e) {
+            acceptGooglePlayTermsOfService();
+            try {
+                updateAppFromMarketPageInternal();
+            } catch (UiObjectNotFoundException ex) {
+                throw new AssertionError(e);
+            }
+        }
+    }
```
Looks like in this entire block, the only thing different from `install` is `line 204` i.e. `updateAppFromMarketPageInternal`. Everything else is same as `install` right? I think there is definitely opportunity for code sharing.

somalaya replied:
Line 204 and 208. Rest is just calling updateAppFromMarketPageInternal and catching exception.

shahzaibj replied:
Yes, so I think definitely opportunity for code sharing here

shahzaibj commented on uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/App.java:
```diff
@@ -116,6 +134,19 @@ public void launch() {
+        String appHint;
+        if (updateAppInstaller instanceof LocalApkInstaller && !TextUtils.isEmpty(localUpdateApkFileName)) {
+            Logger.i(TAG, "Updating the " + this.appName + " from local apk..");
+            appHint = localUpdateApkFileName;
+        } else {
+            Logger.i(TAG, "Updating the " + this.appName + " from Play Store..");
+            appHint = packageName;
+        }
+        updateAppInstaller.updateApp(appHint);
```
```suggestion
            updateAppInstaller.updateApp(localUpdateApkFileName);
        } else {
            Logger.i(TAG, "Updating the " + this.appName + " from Play Store..");
            updateAppInstaller.updateApp(packageName);
        }
```

shahzaibj reviewed (approved):
I left minor comments

somalaya committed:
removed duplicate code

somalaya commented:
> I left minor comments

addressed all comments. Please refer to this commit for playstore code duplication removal changes 6d285f43f549efdfff4579740bee706abf8b399f

somalaya committed:
Merge branch 'dev' into somalaya/SupportMultipleVersionsForAutomation

somalaya committed:
Merge branch 'dev' into somalaya/SupportMultipleVersionsForAutomation
